src/ring_theory/roots_of_unity.lean

```diff
@@ -0,0 +1,586 @@
+and if `l` satisfies `ζ ^ l = 1` then `k ∣ l`. -/
+structure is_primitive_root (ζ : M) (k : ℕ) : Prop :=
+(pow_eq_one : ζ ^ (k : ℕ) = 1)
+(dvd_of_pow_eq_one : ∀ l : ℕ, ζ ^ l = 1 → k ∣ l)
+
+section primitive_roots
+variables {k : ℕ}
+
+/-- `primitive_roots k R` is the finset of primitive `k`-th roots of unity
+in the integral domain `R`. -/
+def primitive_roots (k : ℕ) (R : Type*) [integral_domain R] : finset R :=
+(nth_roots k (1 : R)).to_finset.filter (λ ζ, is_primitive_root ζ k)
+
+@[simp] lemma mem_primitive_roots {ζ : R} (h0 : 0 < k) :
+  ζ ∈ primitive_roots k R ↔ is_primitive_root ζ k :=
+begin
+  rw [primitive_roots, mem_filter, multiset.mem_to_finset, mem_nth_roots h0, and_iff_right_iff_imp],
+  exact is_primitive_root.pow_eq_one
+end
+
+end primitive_roots
+
+namespace is_primitive_root
+
+variables {k l : ℕ}
+
+lemma iff_def (ζ : M) (k : ℕ) :
+  is_primitive_root ζ k ↔ (ζ ^ k = 1) ∧ (∀ l : ℕ, ζ ^ l = 1 → k ∣ l) :=
+⟨λ ⟨h1, h2⟩, ⟨h1, h2⟩, λ ⟨h1, h2⟩, ⟨h1, h2⟩⟩
+
+lemma mk_of_lt (ζ : M) (hk : 0 < k) (h1 : ζ ^ k = 1) (h : ∀ l : ℕ, 0 < l →  l < k → ζ ^ l ≠ 1) :
+  is_primitive_root ζ k :=
+begin
+  refine ⟨h1, _⟩,
+  intros l hl,
+  apply dvd_trans _ (k.gcd_dvd_right l),
+  suffices : k.gcd l = k, { rw this },
+  rw eq_iff_le_not_lt,
+  refine ⟨nat.le_of_dvd hk (k.gcd_dvd_left l), _⟩,
+  intro h', apply h _ (nat.gcd_pos_of_pos_left _ hk) h',
+  exact pow_gcd_eq_one _ h1 hl
+end
+
+section comm_monoid
+
+variables {ζ : M} (h : is_primitive_root ζ k)
+
+lemma pow_eq_one_iff_dvd (l : ℕ) : ζ ^ l = 1 ↔ k ∣ l :=
+⟨h.dvd_of_pow_eq_one l,
+by { rintro ⟨i, rfl⟩, simp only [pow_mul, h.pow_eq_one, one_pow, pnat.mul_coe] }⟩
+
+lemma is_unit (h : is_primitive_root ζ k) (h0 : 0 < k) : is_unit ζ :=
+begin
+  apply is_unit_of_mul_eq_one ζ (ζ ^ (k - 1)),
+  rw [← pow_succ, nat.sub_add_cancel h0, h.pow_eq_one]
+end
+
+lemma pow_ne_one_of_pos_of_lt (h0 : 0 < l) (hl : l < k) : ζ ^ l ≠ 1 :=
+mt (nat.le_of_dvd h0 ∘ h.dvd_of_pow_eq_one _) $ not_le_of_lt hl
+
+lemma pow_inj (h : is_primitive_root ζ k) ⦃i j : ℕ⦄ (hi : i < k) (hj : j < k) (H : ζ ^ i = ζ ^ j) :
+  i = j :=
+begin
+  wlog hij : i ≤ j,
+  apply le_antisymm hij,
+  rw ← nat.sub_eq_zero_iff_le,
+  apply nat.eq_zero_of_dvd_of_lt _ (lt_of_le_of_lt (nat.sub_le_self _ _) hj),
+  apply h.dvd_of_pow_eq_one,
+  rw [← ((h.is_unit (lt_of_le_of_lt (nat.zero_le _) hi)).pow i).mul_left_inj,
+      ← pow_add, nat.sub_add_cancel hij, H, one_mul]
+end
+
+lemma one : is_primitive_root (1 : M) 1 :=
+{ pow_eq_one := pow_one _,
+  dvd_of_pow_eq_one := λ l hl, one_dvd _ }
+
+@[simp] lemma one_right_iff : is_primitive_root ζ 1 ↔ ζ = 1 :=
+begin
+  split,
+  { intro h, rw [← pow_one ζ, h.pow_eq_one] },
+  { rintro rfl, exact one }
+end
+
+@[simp] lemma coe_units_iff {ζ : units M} :
+  is_primitive_root (ζ : M) k ↔ is_primitive_root ζ k :=
+by simp only [iff_def, units.ext_iff, units.coe_pow, units.coe_one]
+
+lemma pow_of_coprime (h : is_primitive_root ζ k) (i : ℕ) (hi : i.coprime k) :
```

Comment:
```suggestion
lemma pow_of_coprime (i : ℕ) (hi : i.coprime k) :
```
This is in scope as a variable, right? Also below

Reply:
But it won't be auto include, since it isn't mentioned before `begin`. I considered `include h`, but that thought I would just go for explicit.